File: server/src/services/ollamaService.js
const axios = require('axios');
const { ai: logger } = require('./logger');

// Ollama configuration
const OLLAMA_BASE_URL = process.env.OLLAMA_BASE_URL || 'http://localhost:11434';
const OLLAMA_MODEL = process.env.OLLAMA_MODEL || 'llama3.2:3b';

class OllamaService {
  constructor() {
    this.baseURL = OLLAMA_BASE_URL;
    this.model = OLLAMA_MODEL;
    logger.info(`Using model ${OLLAMA_MODEL} for Luma responses`);
    this.isAvailable = false;
    this.initializeService();
  }

  async initializeService() {
    try {
      await this.checkAvailability();
      logger.info('Ollama service initialized successfully');
    } catch (error) {
      logger.error('Failed to initialize Ollama service:', error);
    }
  }

  async checkAvailability() {
    try {
      const response = await axios.get(`${this.baseURL}/api/tags`, {
        timeout: 5000
      });
      
      const availableModels = response.data.models || [];
      this.isAvailable = availableModels.some(model => 
        model.name.includes('llama3.2') || model.name.includes('llama3.3')
      );
      
      if (this.isAvailable) {
        logger.info(`Ollama is available with model: ${this.model}`);
      } else {
        logger.warn('Llama 3.2/3.3 model not found in Ollama');
      }
      
      return this.isAvailable;
    } catch (error) {
      logger.error('Ollama service unavailable:', error.message);
      this.isAvailable = false;
      return false;
    }
  }

  async generateResponse(prompt, options = {}) {
    if (!this.isAvailable) {
      await this.checkAvailability();
      if (!this.isAvailable) {
        throw new Error('Ollama service is not available');
      }
    }

    try {
      const response = await axios.post(`${this.baseURL}/api/generate`, {
        model: this.model,
        prompt: prompt,
        stream: false,
        options: {
          temperature: options.temperature || 0.3,
          top_p: options.top_p || 0.9,
          max_tokens: options.max_tokens || 1000,
          ...options
        }
      }, {
        timeout: 60000, // 60 second timeout
        headers: {
          'Content-Type': 'application/json'
        }
      });

      if (response.data && response.data.response) {
        logger.info('Ollama response generated successfully');
        return response.data.response.trim();
      } else {
        throw new Error('Invalid response from Ollama');
      }
    } catch (error) {
      logger.error('Error generating Ollama response:', error.message);
      throw error;
    }
  }

  async chatCompletion(messages, options = {}) {
    if (!this.isAvailable) {
      await this.checkAvailability();
      if (!this.isAvailable) {
        throw new Error('Ollama service is not available');
      }
    }

    try {
      // Convert OpenAI-style messages to Ollama prompt format
      const prompt = this.convertMessagesToPrompt(messages);
      
      const response = await axios.post(`${this.baseURL}/api/generate`, {
        model: this.model,
        prompt: prompt,
        stream: false,
        options: {
          temperature: options.temperature || 0.3,
          top_p: options.top_p || 0.9,
          max_tokens: options.max_tokens || 1000,
          ...options
        }
      }, {
        timeout: 60000, // 60 second timeout for complex queries
        headers: {
          'Content-Type': 'application/json'
        }
      });

      if (response.data && response.data.response) {
        logger.info('Ollama chat completion generated successfully');
        return {
          choices: [{
            message: {
              role: 'assistant',
              content: response.data.response.trim()
            }
          }]
        };
      } else {
        throw new Error('Invalid response from Ollama');
      }
    } catch (error) {
      logger.error('Error in Ollama chat completion:', error.message);
      throw error;
    }
  }

  convertMessagesToPrompt(messages) {
    let prompt = '';
    
    messages.forEach(message => {
      switch (message.role) {
        case 'system':
          prompt += `System: ${message.content}\n\n`;
          break;
        case 'user':
          prompt += `Human: ${message.content}\n\n`;
          break;
        case 'assistant':
          prompt += `Assistant: ${message.content}\n\n`;
          break;
      }
    });

    prompt += 'Assistant: ';
    return prompt;
  }

  // Enhanced Medical symptom analysis using Llama 3.2 with multi-modal approach
  async analyzeSymptoms(symptoms, patientContext = {}, additionalData = {}) {
    const { vitals, labResults, recentSymptoms, environmentalFactors } = additionalData;
    
    const prompt = `You are an advanced medical AI assistant specializing in comprehensive symptom analysis and early disease detection.

Patient Context:
- Age: ${patientContext.age || 'Unknown'}
- Gender: ${patientContext.gender || 'Unknown'}
- Medical History: ${JSON.stringify(patientContext.medicalHistory || [])}
- Chronic Conditions: ${JSON.stringify(patientContext.chronicConditions || [])}
- Current Medications: ${JSON.stringify(patientContext.medications || [])}
- Allergies: ${JSON.stringify(patientContext.allergies || [])}
- Risk Factors: ${JSON.stringify(patientContext.riskFactors || [])}

Current Symptoms: ${symptoms}

${vitals ? `Recent Vital Signs:\n${JSON.stringify(vitals, null, 2)}\n` : ''}
${labResults ? `Latest Lab Results:\n${JSON.stringify(labResults, null, 2)}\n` : ''}
${recentSymptoms ? `Recent Symptom History:\n${JSON.stringify(recentSymptoms, null, 2)}\n` : ''}
${environmentalFactors ? `Environmental Context:\n${JSON.stringify(environmentalFactors, null, 2)}\n` : ''}

Please provide a comprehensive analysis in JSON format:
{
  "primaryDiagnosis": "Most likely condition based on all available data",
  "confidence": 0-100,
  "severity": "low|medium|high|critical",
  "urgency": "low|moderate|high|critical",
  "possibleConditions": [
    {
      "condition": "condition name",
      "probability": 0-100,
      "supportingEvidence": ["evidence1", "evidence2"]
    }
  ],
  "differentialDiagnosis": ["alternative conditions to consider"],
  "redFlags": ["serious warning signs present"],
  "recommendations": {
    "immediate": ["urgent actions needed"],
    "shortTerm": ["actions needed within 24-48 hours"],
    "longTerm": ["ongoing management recommendations"]
  },
  "requiredTests": ["specific diagnostic tests recommended"],
  "followUpRecommendations": ["follow-up care suggestions"],
  "riskFactorAnalysis": {
    "modifiable": ["risk factors that can be changed"],
    "nonModifiable": ["risk factors that cannot be changed"]
  },
  "preventiveActions": ["actions to prevent progression"],
  "monitoringParameters": ["symptoms/vitals to monitor"]
}

Analyze patterns, consider seasonal factors, and provide evidence-based recommendations. Always prioritize patient safety. Respond only with valid JSON.`;

    try {
      const response = await this.generateResponse(prompt, {
        temperature: 0.3,
        max_tokens: 1000
      });

      // Try to parse JSON response
      let analysis;
      try {
        analysis = JSON.parse(response);
      } catch (parseError) {
        logger.warn('Failed to parse JSON response, extracting relevant information');
        analysis = this.extractAnalysisFromText(response, symptoms);
      }

      // Validate and ensure required fields
      return this.validateAnalysis(analysis);
    } catch (error) {
      logger.error('Error in Ollama symptom analysis:', error.message);
      throw error;
    }
  }

  // Health risk assessment using Llama 3.2
  async assessHealthRisk(patientData) {
    const prompt = `You are a healthcare AI assistant. Assess the health risk for this patient:

Patient Data:
${JSON.stringify(patientData, null, 2)}

Please provide a comprehensive health risk assessment in JSON format:
{
  "overallRisk": "low|medium|high|critical",
  "riskFactors": ["factor1", "factor2", ...],
  "recommendations": ["recommendation1", "recommendation2", ...],
  "preventiveMeasures": ["measure1", "measure2", ...],
  "monitoringRequired": ["item1", "item2", ...]
}

Focus on preventive care and early intervention. Respond only with valid JSON.`;

    try {
      const response = await this.generateResponse(prompt, {
        temperature: 0.2,
        max_tokens: 800
      });

      let assessment;
      try {
        assessment = JSON.parse(response);
      } catch (parseError) {
        logger.warn('Failed to parse risk assessment JSON');
        assessment = {
          overallRisk: 'medium',
          riskFactors: ['Assessment requires manual review'],
          recommendations: ['Consult with healthcare provider'],
          preventiveMeasures: ['Regular health check-ups'],
          monitoringRequired: ['General health monitoring']
        };
      }

      return assessment;
    } catch (error) {
      logger.error('Error in Ollama health risk assessment:', error.message);
      throw error;
    }
  }

  // Luma chatbot responses using Llama 3.2
  async getLumaResponse(query, context = {}) {
    const prompt = `You are Luma, a healthcare AI assistant. You provide helpful, accurate medical guidance while always recommending professional medical consultation for serious issues.

User asked: "${query}"

Please provide a helpful, clear response that:
- Is medically accurate and evidence-based
- Uses simple, understandable language
- Recommends professional care for serious symptoms
- If emergency symptoms (chest pain, difficulty breathing, severe bleeding), start with "🚨 URGENT: Seek immediate medical attention."

Response:`;

    try {
      const response = await this.generateResponse(prompt, {
        temperature: 0.2, // more deterministic
        max_tokens: 800 // allowance for deeper perspective
      });

      // Clean up the response
      let cleanedResponse = response.trim();
      
      // Remove any unwanted prefixes that might be generated
      cleanedResponse = cleanedResponse.replace(/^(Response:|Assistant:|Luma:)\s*/i, '');
      
      // Ensure the response ends with a professional disclaimer if it's medical advice
      const medicalKeywords = ['symptom', 'diagnosis', 'treatment', 'medication', 'condition', 'disease', 'pain', 'illness'];
      const containsMedicalAdvice = medicalKeywords.some(keyword => 
        query.toLowerCase().includes(keyword) || cleanedResponse.toLowerCase().includes(keyword)
      );
      
      if (containsMedicalAdvice && !cleanedResponse.toLowerCase().includes('healthcare provider') && !cleanedResponse.toLowerCase().includes('medical professional')) {
        cleanedResponse += '\n\nPlease remember to consult with a healthcare provider for proper medical evaluation and treatment.';
      }

      return {
        response: cleanedResponse,
        type: 'general',
        source: 'llama3.2'
      };
    } catch (error) {
      logger.error('Error in Luma response generation:', error.message);
      throw error;
    }
  }

  // Predict health trends using Llama 3.2
  async predictHealthTrends(patientHistory) {
    const prompt = `You are a predictive health AI. Analyze this patient's health history and predict potential trends:

Patient History:
${JSON.stringify(patientHistory, null, 2)}

Provide predictions in JSON format:
{
  "riskProgression": "improving|stable|declining|critical",
  "predictedConditions": ["condition1", "condition2", ...],
  "recommendations": ["recommendation1", "recommendation2", ...],
  "timeframe": "short-term|medium-term|long-term",
  "confidence": 0-100,
  "keyIndicators": ["indicator1", "indicator2", ...]
}

Focus on preventive care and early intervention opportunities. Respond only with valid JSON.`;

    try {
      const response = await this.generateResponse(prompt, {
        temperature: 0.3,
        max_tokens: 800
      });

      let trends;
      try {
        trends = JSON.parse(response);
      } catch (parseError) {
        logger.warn('Failed to parse trends JSON');
        trends = {
          riskProgression: 'stable',
          predictedConditions: [],
          recommendations: ['Regular health monitoring recommended'],
          timeframe: 'medium-term',
          confidence: 50,
          keyIndicators: ['Overall health status']
        };
      }

      return trends;
    } catch (error) {
      logger.error('Error in health trend prediction:', error.message);
      throw error;
    }
  }

  extractAnalysisFromText(text, symptoms) {
    // Fallback method to extract key information if JSON parsing fails
    return {
      primaryDiagnosis: 'Symptoms require medical evaluation',
      severity: 'medium',
      confidence: 60,
      possibleConditions: [symptoms.split(' ')[0] + ' related condition'],
      recommendations: ['Consult with healthcare provider', 'Monitor symptoms'],
      urgency: 'moderate',
      redFlags: [],
      followUpRecommendations: ['Follow up if symptoms persist']
    };
  }

  validateAnalysis(analysis) {
    // Ensure all required fields are present
    return {
      primaryDiagnosis: analysis.primaryDiagnosis || 'Symptoms require evaluation',
      severity: analysis.severity || 'medium',
      confidence: Math.min(100, Math.max(0, analysis.confidence || 60)),
      possibleConditions: Array.isArray(analysis.possibleConditions) ? analysis.possibleConditions : [],
      recommendations: Array.isArray(analysis.recommendations) ? analysis.recommendations : ['Consult healthcare provider'],
      urgency: analysis.urgency || 'moderate',
      redFlags: Array.isArray(analysis.redFlags) ? analysis.redFlags : [],
      followUpRecommendations: Array.isArray(analysis.followUpRecommendations) ? analysis.followUpRecommendations : []
    };
  }
}

module.exports = new OllamaService();

File: python-ai/services/ollama_service.py
"""
Ollama Service for BioVerse
Handles local AI model interactions through Ollama
"""

import asyncio
import json
import httpx
from typing import Dict, List, Optional, Any
from pydantic import BaseModel
import os
from .base_service import BaseService

class OllamaRequest(BaseModel):
    model: str
    prompt: str
    stream: bool = False
    options: Optional[Dict[str, Any]] = None

class OllamaResponse(BaseModel):
    model: str
    response: str
    done: bool
    context: Optional[List[int]] = None

class OllamaService(BaseService):
    """Service for interacting with Ollama local AI models"""
    
    def __init__(self):
        super().__init__("OllamaService")
        self.base_url = os.getenv("OLLAMA_BASE_URL", "http://localhost:11434")
        self.default_model = os.getenv("OLLAMA_MODEL", "deepseek-r1:1.5b")
        self.embedding_model = os.getenv("OLLAMA_EMBEDDING_MODEL", "nomic-embed-text")
        self.vision_model = os.getenv("OLLAMA_VISION_MODEL", "llava:7b")
        self.client = None
        self.is_available = False
        self.available_models = []
        
    async def initialize(self):
        """Initialize the Ollama service"""
        try:
            self.client = httpx.AsyncClient(timeout=30.0)
            
            # Check if Ollama is running
            await self._check_availability()
            
            if self.is_available:
                # Get available models
                await self._load_available_models()
                
                # Ensure required models are available
                await self._ensure_models()
                
            self.logger.info(f"Ollama service initialized. Available: {self.is_available}")
            
        except Exception as e:
            self.logger.error(f"Failed to initialize Ollama service: {e}")
            self.is_available = False
    
    async def _check_availability(self):
        """Check if Ollama service is available"""
        try:
            response = await self.client.get(f"{self.base_url}/api/tags")
            self.is_available = response.status_code == 200
        except Exception as e:
            self.logger.warning(f"Ollama not available: {e}")
            self.is_available = False
    
    async def _load_available_models(self):
        """Load list of available models"""
        try:
            response = await self.client.get(f"{self.base_url}/api/tags")
            if response.status_code == 200:
                data = response.json()
                self.available_models = [model["name"] for model in data.get("models", [])]
                self.logger.info(f"Available models: {self.available_models}")
        except Exception as e:
            self.logger.error(f"Failed to load available models: {e}")
    
    async def _ensure_models(self):
        """Ensure required models are available"""
        required_models = [self.default_model, self.embedding_model]
        
        for model in required_models:
            if model not in self.available_models:
                self.logger.info(f"Pulling required model: {model}")
                await self._pull_model(model)
    
    async def _pull_model(self, model_name: str):
        """Pull a model from Ollama registry"""
        try:
            response = await self.client.post(
                f"{self.base_url}/api/pull",
                json={"name": model_name},
                timeout=300.0  # 5 minutes for model download
            )
            
            if response.status_code == 200:
                self.logger.info(f"Successfully pulled model: {model_name}")
                self.available_models.append(model_name)
            else:
                self.logger.error(f"Failed to pull model {model_name}: {response.text}")
                
        except Exception as e:
            self.logger.error(f"Error pulling model {model_name}: {e}")
    
    async def generate_text(self, prompt: str, model: Optional[str] = None, **kwargs) -> str:
        """Generate text using Ollama model"""
        if not self.is_available:
            raise Exception("Ollama service not available")
        
        model = model or self.default_model
        
        try:
            request_data = {
                "model": model,
                "prompt": prompt,
                "stream": False,
                "options": kwargs
            }
            
            response = await self.client.post(
                f"{self.base_url}/api/generate",
                json=request_data,
                timeout=60.0
            )
            
            if response.status_code == 200:
                result = response.json()
                return result.get("response", "")
            else:
                raise Exception(f"Ollama API error: {response.status_code} - {response.text}")
                
        except Exception as e:
            self.logger.error(f"Error generating text: {e}")
            raise
    
    async def generate_health_analysis(self, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        """Generate health analysis using AI"""
        prompt = f"""
        As a medical AI assistant, analyze the following patient data and provide insights:
        
        Patient Data:
        {json.dumps(patient_data, indent=2)}
        
        Please provide:
        1. Health risk assessment
        2. Potential concerns
        3. Recommendations
        4. Preventive measures
        
        Format your response as JSON with the following structure:
        {{
            "risk_level": "low|medium|high",
            "risk_score": 0-100,
            "concerns": ["concern1", "concern2"],
            "recommendations": ["rec1", "rec2"],
            "preventive_measures": ["measure1", "measure2"],
            "summary": "Brief summary"
        }}
        """
        
        try:
            response = await self.generate_text(prompt, temperature=0.3, top_p=0.9)
            
            # Try to parse JSON response
            try:
                return json.loads(response)
            except json.JSONDecodeError:
                # If not valid JSON, return structured response
                return {
                    "risk_level": "medium",
                    "risk_score": 50,
                    "concerns": ["Unable to parse detailed analysis"],
                    "recommendations": ["Consult healthcare provider"],
                    "preventive_measures": ["Regular health checkups"],
                    "summary": response[:200] + "..." if len(response) > 200 else response
                }
                
        except Exception as e:
            self.logger.error(f"Error generating health analysis: {e}")
            raise
    
    async def generate_embeddings(self, text: str, model: Optional[str] = None) -> List[float]:
        """Generate embeddings for text"""
        if not self.is_available:
            raise Exception("Ollama service not available")
        
        model = model or self.embedding_model
        
        try:
            response = await self.client.post(
                f"{self.base_url}/api/embeddings",
                json={
                    "model": model,
                    "prompt": text
                },
                timeout=30.0
            )
            
            if response.status_code == 200:
                result = response.json()
                return result.get("embedding", [])
            else:
                raise Exception(f"Ollama embeddings error: {response.status_code}")
                
        except Exception as e:
            self.logger.error(f"Error generating embeddings: {e}")
            raise
    
    async def analyze_symptoms(self, symptoms: List[str]) -> Dict[str, Any]:
        """Analyze symptoms and provide potential diagnoses"""
        symptoms_text = ", ".join(symptoms)
        
        prompt = f"""
        Analyze these symptoms and provide potential medical insights:
        Symptoms: {symptoms_text}
        
        Provide a JSON response with:
        {{
            "potential_conditions": ["condition1", "condition2"],
            "urgency_level": "low|medium|high|emergency",
            "recommended_actions": ["action1", "action2"],
            "specialist_referral": "specialist_type or null",
            "confidence": 0-100
        }}
        
        Important: This is for informational purposes only and not a substitute for professional medical advice.
        """
        
        try:
            response = await self.generate_text(prompt, temperature=0.2)
            
            try:
                return json.loads(response)
            except json.JSONDecodeError:
                return {
                    "potential_conditions": ["Analysis unavailable"],
                    "urgency_level": "medium",
                    "recommended_actions": ["Consult healthcare provider"],
                    "specialist_referral": None,
                    "confidence": 50
                }
                
        except Exception as e:
            self.logger.error(f"Error analyzing symptoms: {e}")
            raise
    
    async def generate_health_recommendations(self, health_profile: Dict[str, Any]) -> List[str]:
        """Generate personalized health recommendations"""
        prompt = f"""
        Based on this health profile, provide 5 personalized health recommendations:
        
        Health Profile:
        {json.dumps(health_profile, indent=2)}
        
        Provide practical, actionable recommendations as a JSON array:
        ["recommendation1", "recommendation2", "recommendation3", "recommendation4", "recommendation5"]
        """
        
        try:
            response = await self.generate_text(prompt, temperature=0.4)
            
            try:
                return json.loads(response)
            except json.JSONDecodeError:
                # Return default recommendations if parsing fails
                return [
                    "Maintain regular exercise routine",
                    "Follow a balanced diet",
                    "Get adequate sleep (7-9 hours)",
                    "Stay hydrated",
                    "Schedule regular health checkups"
                ]
                
        except Exception as e:
            self.logger.error(f"Error generating recommendations: {e}")
            raise
    
    async def luma_chat(self, message: str, context: Optional[Dict] = None) -> Dict[str, Any]:
        """Luma AI health assistant chat using Ollama"""
        if not self.is_available:
            return {
                "response": f"Hello! I'm Luma, your AI health assistant. You asked: '{message}'. I'm currently running in offline mode, but I can still provide basic health guidance. Always consult healthcare professionals for medical advice.",
                "confidence": 0.7,
                "model_used": "fallback"
            }
        
        try:
            luma_prompt = f"""You are Luma, BioVerse's friendly health AI assistant. Provide helpful, accurate health information while being empathetic. Always recommend consulting healthcare professionals for serious concerns.

User asked: "{message}"

Respond as Luma with helpful health guidance."""

            response_text = await self.generate_text(
                luma_prompt,
                temperature=0.7,
                top_p=0.9
            )
            
            return {
                "response": response_text.strip(),
                "confidence": 0.9,
                "model_used": self.default_model
            }
            
        except Exception as e:
            self.logger.error(f"Error in Luma chat: {e}")
            return {
                "response": f"I'm sorry, I'm having trouble right now. Regarding '{message}', I'd recommend consulting with a healthcare professional.",
                "confidence": 0.5,
                "model_used": "error_fallback"
            }

    async def close(self):
        """Close the Ollama service"""
        if self.client:
            await self.client.aclose()
        self.logger.info("Ollama service closed")